include/simdjson/padded_string.h

```diff
@@ -33,77 +46,156 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
+    // free the payload
+    simdjson::aligned_free_char(other_.data_ptr);
+    other_.data_ptr = nullptr;
+    other_.viable_size = 0;
+    return *this;
   }
 
-  padded_string(padded_string &&o) noexcept
-      : viable_size(o.viable_size), data_ptr(o.data_ptr) {
-    o.data_ptr = nullptr; // we take ownership
+  // load from file to padded_string
+  // reset the ps_
+  // return the result of fread()
+  // return 0 on error
+  static errno_t load(padded_string &ps_, FILE *fp) noexcept {
+
+    if (std::fseek(fp, 0, SEEK_END) < 0) {
+      std::fclose(fp);
+      return errno;
+    }
+    long llen = std::ftell(fp);
+    if ((llen < 0) || (llen == LONG_MAX)) {
+      std::fclose(fp);
+      return errno;
+    }
+
+    // now make the temporary p string
+    size_t length_ = (size_t)llen;
+
+    if (!ps_.reset(length_))
+      return ENOMEM;
+
+    std::rewind(fp);
+    size_t readb = std::fread(ps_.data(), 1, length_, fp);
+    std::fclose(fp);
+
+    if (readb != length_) {
+      // free mem before bad return
+      ps_.~padded_string();
+      return errno;
+    }
+    return 0;
   }
 
-  padded_string &operator=(padded_string &&o) {
-    data_ptr = o.data_ptr;
-    viable_size = o.viable_size;
-    o.data_ptr = nullptr; // we take ownership
-    o.viable_size = 0;
-    return *this;
+  /*
+  Makes padded_string 'Swappable'. This implementation is also ADL friendly.
+
+  using namespace simjson ;
+  padded_string a("A");
+  padded_string b("B");
+    swap(a,b) ;
+      printf("%s", a.data() ); // "B"
+  */
+  friend void swap(padded_string &left_, padded_string &right_) {
+
+    using std::swap; // bring in swap for built-in types
+
+    swap(left_.viable_size, right_.viable_size);
+    swap(left_.data_ptr, right_.data_ptr);
   }
 
-  void swap(padded_string &o) {
-    size_t tmp_viable_size = viable_size;
-    char *tmp_data_ptr = data_ptr;
-    viable_size = o.viable_size;
-    data_ptr = o.data_ptr;
-    o.data_ptr = tmp_data_ptr;
-    o.viable_size = tmp_viable_size;
+  /*
+  padded_string a("A");
+  padded_string b("B");
+       a.swap(b) ;
+         printf("%s", a.data() ); // "B"
+  */
+  void swap( padded_string &right_) {
+    // auto &left_ = *this;
+    std::swap(*this, right_);
   }
 
+  /*
+     padded_string a("A");
+     a.~padded_string() ;
+    */
   ~padded_string() {
+    if (data_ptr != nullptr ) {
       aligned_free_char(data_ptr);
+      data_ptr = nullptr;
+      viable_size = size_t(0);
+    }
   }
 
-  size_t size() const  { return viable_size; }
+  size_t size() const { return viable_size; }
 
-  size_t length() const  { return viable_size; }
+  size_t length() const { return viable_size; }
 
-  char *data() const  { return data_ptr; }
+  char *data() const { return data_ptr; }
 
 private:
+    // no copying
   padded_string &operator=(const padded_string &o) = delete;
   padded_string(const padded_string &o) = delete;
 
-  size_t viable_size;
+  size_t viable_size{};
```

Comment:
This looks like an unnecessary change.

Reply:
we have been through this before :)  C++17 standard says if value init aka '{}' is not used variable/constant is initialized to indeterminate value and if opposite it is "zeroed" to "zero" as appropriate to the type in use.  And that is indeed the case for CL (aka MSVC).  

```cpp
// var is "zeroed" to "zero" for the type T
T var{} ;
```

Reply:
I know the C++ syntax, but isn't viable_size initialize in the constructor?

I should be able to look at this line and understand why it was changed. As it is, I have no idea why you changed it.

By itself, I do not find it objectionable, but you are making many, many changes (hundreds of lines) and I am struggling to find the motivation for most of these changes. This makes it hard to review.

Please change as few lines as possible. 

Can you reorg. your PRs so that they are narrow and focused, with clearly stated benefits?

Reply:
I am aware of that and will change that behaviour, so we do not indulge in a ping-pong around each PR :) I might do the issue first, discuss it and then do the PR on the agreed outcome of the issue discussion.